File: clients/simpleAgent.js
const WebSocket = require("ws");

// Configuration
const CONFIG = {
  serverIP: "localhost",
  serverPort: 3000,
  defaultId: "k",
  validDirections: ["up", "down", "left", "right"],
};

// Game state
const gameState = {
  agentId: process.argv[2] || CONFIG.defaultId,
  playerName: null,
};

// Connect to WebSocket server
const ws = new WebSocket(
  `ws://${CONFIG.serverIP}:${CONFIG.serverPort}?id=${gameState.agentId}`
);

ws.on("open", () => console.log("Connected to WebSocket server"));
ws.on("error", (error) => console.error("WebSocket error:", error));
ws.on("close", () => console.log("Disconnected from WebSocket server"));

ws.on("message", (data) => {
  const receivedMsg = JSON.parse(data.toString("utf-8"));

  // Store player name when receiving the connection success message
  if (receivedMsg.message === "Player connected successfully.") {
    console.log(
      `Agent connected with name: '${receivedMsg.name}' and id: '${gameState.agentId}'.`
    );
    gameState.playerName = receivedMsg.name;
    return;
  }

  // Only make a move if the game isn't over
  if (receivedMsg.winner === null || receivedMsg.winner === undefined) {
    // Make a random move
    const move = {
      playerId: gameState.agentId,
      direction:
        CONFIG.validDirections[
          Math.floor(Math.random() * CONFIG.validDirections.length)
        ],
    };

    // Send the move
    ws.send(JSON.stringify(move));
    console.log("Sent move:", move);
  }
});

File: clients/simpleAgent.py
import asyncio
import websockets
import json
import sys
import random

# Configuration
CONFIG = {
    "serverIP": "localhost",
    "serverPort": 3000,
    "defaultId": "k",
    "validDirections": ["up", "down", "left", "right"],
}

# Game state
class GameState:
    def __init__(self):
        self.agentId = sys.argv[1] if len(sys.argv) > 1 else CONFIG["defaultId"]
        self.playerName = None

async def connect_to_game_server(game_state):
    uri = f"ws://{CONFIG['serverIP']}:{CONFIG['serverPort']}?id={game_state.agentId}"

    async with websockets.connect(uri) as websocket:
        print("Connected to WebSocket server")

        while True:
            try:
                data = await websocket.recv()
                received_msg = json.loads(data)

                # Store player name when receiving the connection success message
                if received_msg.get("message") == "Player connected successfully.":
                    print(
                        f"Agent connected with name: '{received_msg['name']}' "
                        f"and id: '{game_state.agentId}'."
                    )
                    game_state.playerName = received_msg["name"]
                    continue

                game_is_over = received_msg.get("winner") is not None

                if not game_is_over and received_msg.get("map"):
                    # Make a random move
                    move = {
                        "playerId": game_state.agentId,
                        "direction": random.choice(CONFIG["validDirections"])
                    }

                    await websocket.send(json.dumps(move))
                    print(f"Sent move: {move}")

            except websockets.exceptions.ConnectionClosed:
                print("Disconnected from WebSocket server")
                break
            except Exception as e:
                print(f"Error: {e}")
                break

def main():
    game_state = GameState()
    print(f"Starting agent with ID: {game_state.agentId}")
    asyncio.run(connect_to_game_server(game_state))

if __name__ == "__main__":
    main()